src/ray/core_worker/transport/direct_actor_transport.h

```diff
@@ -0,0 +1,131 @@
+
+namespace ray {
+
+/// In direct actor call task submitter and receiver, a task is directly submitted
+/// to the actor that will execute it.
+
+struct PendingTaskRequest {
+  PendingTaskRequest(const TaskID &task_id, int num_returns,
+                     std::unique_ptr<rpc::PushTaskRequest> request)
+      : task_id_(task_id), num_returns_(num_returns), request_(std::move(request)) {}
+
+  TaskID task_id_;
+  int num_returns_;
+  std::unique_ptr<rpc::PushTaskRequest> request_;
+};
+
+class CoreWorkerDirectActorTaskSubmitter : public CoreWorkerTaskSubmitter {
+ public:
+  CoreWorkerDirectActorTaskSubmitter(boost::asio::io_service &io_service,
+                                     gcs::RedisGcsClient &gcs_client,
+                                     CoreWorkerObjectInterface &object_interface);
+
+  /// Submit a task to an actor for execution.
+  ///
+  /// \param[in] task The task spec to submit.
+  /// \return Status.
+  Status SubmitTask(const TaskSpecification &task_spec) override;
+
+ private:
+  /// Subscribe for all actor updates.
+  Status SubscribeActorUpdates();
+
+  /// Helper function to push a task to an actor.
+  ///
+  /// \param[in] client The RPC client to send tasks to an actor.
+  /// \param[in] request The request to send.
+  /// \param[in] task_id The ID of a task.
+  /// \param[in] num_returns Number of return objects.
+  /// \return Status.
+  Status PushTask(rpc::DirectActorClient &client, const rpc::PushTaskRequest &request,
+                  const TaskID &task_id, int num_returns);
+
+  /// Treat a task as failed.
+  ///
+  /// \param[in] task_id The ID of a task.
+  /// \param[in] num_returns Number of return objects.
+  /// \param[in] error_type The type of the specific error.
+  /// \return Void.
+  void TreatTaskAsFailed(const TaskID &task_id, int num_returns,
+                         const rpc::ErrorType &error_type);
+
+  /// Handle the case that an actor becomes alive.
+  /// Note that this function doesn't take lock, the caller is expected to hold
+  /// `rpc_clients_mutex_` before calling this function.
+  ///
+  /// \param[in] actor_id Actor ID.
+  /// \param[in] ip_address The ip address of the node that the actor is running on.
+  /// \param[in] port The port that the actor is listening on.
+  /// \return Void.
+  void HandleActorAlive(const ActorID &actor_id, std::string ip_address, int port);
+
+  /// The IO event loop.
+  boost::asio::io_service &io_service_;
+
+  /// Gcs client.
+  gcs::RedisGcsClient &gcs_client_;
+
+  /// The `ClientCallManager` object that is shared by all `DirectActorClient`s.
+  rpc::ClientCallManager client_call_manager_;
+
+  /// Mutex to proect `rpc_clients_` below.
+  std::mutex rpc_clients_mutex_;
+
+  /// Map from actor ids to direct actor call rpc clients.
+  std::unordered_map<ActorID, std::unique_ptr<rpc::DirectActorClient>> rpc_clients_;
+
+  /// Map from actor ids to actor's state.
+  std::unordered_map<ActorID, gcs::ActorTableData::ActorState> actor_states_;
```

Comment:
The size of these maps grows forever, is that right? At what point can we garbage collect these entries?

Reply:
This is currently not garbage collected, the map will get destroyed when this worker dies, this is similar to the `actor_registry_` in raylet. Later when the support to subscribe for a specific actor is ready, we can limit this map to only contain entries for actors that we submit tasks to. I will add a TODO here.

Reply:
Sounds good. Let's make sure to prioritize it since it will become a performance / stability issue.

Reply:
sure, thanks.